armi/reactor/components/component.py

```diff
@@ -734,22 +752,71 @@ def setNumberDensities(self, numberDensities):
         """
-        self.p.numberDensities = numberDensities
+        self.updateNumberDensities(numberDensities, wipe=True)
 
-    def updateNumberDensities(self, numberDensities):
+    def updateNumberDensities(self, numberDensities, wipe=False):
         """
         Set one or more multiple number densities. Leaves unlisted number densities alone.
 
         Parameters
         ----------
         numberDensities : dict
             nucName: ndens pairs.
+        wipe : bool, optional
+            Controls whether the old number densities are wiped. Any nuclide densities not
+            provided in numberDensities will be effectively set to 0.0.
+
+        Notes
+        -----
+        Sometimes volume/dimensions change due to number density change when the material thermal
+        expansion depends on the component's composition (e.g. its plutonium fraction). In this
+        case, changing the density will implicitly change the area/volume. Since it is difficult to
+        predict the new dimensions, and perturbation/depletion calculations almost exclusively
+        assume constant volume, the densities sent are automatically adjusted to conserve mass with
+        the original dimensions. That is, the component's densities are not exactly as passed, but
+        whatever they would need to be to preserve volume integrated number densities (moles) from
+        the pre-perturbed component's volume/dimensions.
+
+        This has no effect if the material thermal expansion has no dependence on component
+        composition. If this is not desired, `self.p.numberDensities` can be set directly.
         """
+        # prepare to change the densities with knowledge that dims could change due to
+        # material thermal expansion dependence on composition
+        if len(self.p.numberDensities) > 0:
+            dLLprev = (
+                self.material.linearExpansionPercent(Tc=self.temperatureInC) / 100.0
+            )
+            materialExpansion = True
+        else:
+            dLLprev = 0.0
+            materialExpansion = False
+
+        try:
+            vol = self.getVolume()
+        except (AttributeError, TypeError):
+            # either no parent to get height or parent's height is None
+            # which would be AttributeError and TypeError respectively, but other errors could be possible
+            vol = None
+            area = self.getArea()
+
+        # change the densities
+        if wipe:
+            self.p.numberDensities = {}  # clear things not passed
         self.p.numberDensities.update(numberDensities)
-        # since we're updating the object the param points to but not the param itself, we have to inform
-        # the param system to flag it as modified so it properly syncs during ``syncMpiState``.
-        self.p.assigned = parameters.SINCE_ANYTHING
-        self.p.paramDefs["numberDensities"].assigned = parameters.SINCE_ANYTHING
```

Comment:
We can just remove this without hurting parallel runs? That's surprising.

Reply:
Nope, this caused a bug. reverting now...

Reply:
Ah, okay, that makes more sense.

Thanks!